vignettes/migrating.Rmd

```diff
@@ -0,0 +1,112 @@
+---
+title: "Migrating from dust 1.x.x"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Migrating from dust 1.x.x}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+This vignette discusses key differences with dust version 1.  Version 2 of dust is a full rewrite, and some features have moved out of the package, some have moved in from `mcstate` and others have not yet been implemented.
+
+If you have not used previous versions of `dust` and `mcstate` (e.g., via `odin.dust`), you do not need to read this vignette.
+
+```{r}
+library(dust2)
+```
+
+# Dude, where's my feature?
+
+Features now found elsewhere:
+
+* The random number generation support is now found in [`monty`](https://mrc-ide.github.io/monty); this affects very few users.
+
+Features now included:
+
+* The particle filter is now fully implemented in dust, and no longer part of `mcstate` (now `monty`)
+
+Not yet implemented:
+
+* GPU compilation (a rewrite is planned for 2025)
+* "Restart" a particle filter
+* Multi-phase systems (not believed used except [sircovid](https://mrc-ide.github.io/sircovid))
+
+# New interface
+
+The new interface is completely different, following feedback about the discoverability and documentation of features within the package.  We've also generally tried to make the functions more self-contained, less magic, and easier to understand.  Some lesser-used functions have been removed entirely.  We apologise for the inconvenience that this will cause in migrating old code, but unfortunately this cannot be done automatically.
+
+Previously, dust generators were "[R6 classes](https://r6.r-lib.org/)", which were created using `$new()` and then interacted with using methods (e.g., `sys$run()`.  In contrast, `dust2` uses free functions to create systems and to interact with these systems once created.
+
+In this table, consider the dust system `sir`, which is available via `dust_example()` in both versions of the package.  In dust 1.0.0 could allocate a system with 10 particles, set the initial conditions, run it forward for 20 time units and return the state matrix by writing:
+
+```r
+sir <- dust_example("sir")
+sys <- sir$new(pars = list(dt = 1), time = 0, n_particles = 10)
+sys$run(10)
+```
+
+And in dust2 we could write
+
+```{r}
+sir <- dust_example("sir")
+sys <- dust_system_create(sir, pars = list(), time = 0, n_particles = 10)
+dust_system_set_state_initial(sys)
+dust_system_run_to_time(sys, 10)
+dust_system_state(sys)
+```
```

Comment:
`dt = 1` is explicit in dust 1, and doesn't occur in the migrated version as it's the default - do we need to spell that out?

Reply:
thanks - I think this needs a whole section, I'll add one today